File: codes/python/gemini/minimal/task_3/Solution.py
import json
import heapq

def calculate_travel_time(distance, region_type, congestion_factor):
    if region_type == "Urban":
        base_speed = 30
    elif region_type == "Suburban":
        base_speed = 50
    else:  # Rural
        base_speed = 80
    
    travel_time = (distance / (base_speed * congestion_factor)) * 60
    return travel_time

def build_graph(regions):
    graph = {}
    for region in regions:
        graph[region["region_id"]] = {}
        for neighbor, distance in region["distances"].items():
            if neighbor not in graph:
                continue
            travel_time = calculate_travel_time(distance, region["type"], region["congestion_factor"])

            if neighbor in region["blocked_routes"]:
                travel_time = float('inf')
            
            graph[region["region_id"]][neighbor] = travel_time
            
    return graph

def dijkstra(graph, start, end, available_nodes):
    distances = {node: float('inf') for node in graph}
    distances[start] = 0
    priority_queue = [(0, start)]  # (distance, node)
    
    while priority_queue:
        current_distance, current_node = heapq.heappop(priority_queue)
        
        if current_distance > distances[current_node]:
            continue
        
        if current_node == end:
            return distances[end]
        
        for neighbor, weight in graph[current_node].items():
            if neighbor not in available_nodes:
                continue
            distance = current_distance + weight
            if distance < distances[neighbor]:
                distances[neighbor] = distance
                heapq.heappush(priority_queue, (distance, neighbor))
    
    return float('inf')

def dispatch_ambulance(emergency_event, ambulances, hospitals, regions):
    emergency_id = emergency_event["id"]
    event_region = emergency_event["region_id"]
    severity = emergency_event["severity"]
    emergency_type = emergency_event["type"]
    special_requirements = emergency_event["special_requirements"]

    graph = build_graph(regions)
    available_regions = list(graph.keys())

    # Step 1: Select the fastest ambulance
    best_ambulance = None
    min_ambulance_time = float('inf')

    for ambulance in ambulances:
        if ambulance["status"] == "Available":
            ambulance_id = ambulance["id"]
            ambulance_region = ambulance["current_region"]
            ambulance_capabilities = ambulance["capabilities"]

            # Check capability requirements for critical emergencies
            if severity == "Critical" and "Advanced Life Support" in special_requirements and "Advanced Life Support" not in ambulance_capabilities and "Defibrillator" not in ambulance_capabilities:
                continue  # Skip if ambulance lacks required capabilities

            ambulance_time = dijkstra(graph, ambulance_region, event_region, available_regions)

            if ambulance_time < min_ambulance_time:
                min_ambulance_time = ambulance_time
                best_ambulance = ambulance
            elif ambulance_time == min_ambulance_time and best_ambulance is not None:
                # Prioritize by capabilities, then lower ID
                best_ambulance_capabilities = best_ambulance["capabilities"]
                if severity == "Critical" and "Advanced Life Support" in special_requirements:
                    if "Advanced Life Support" in ambulance_capabilities and "Advanced Life Support" not in best_ambulance_capabilities:
                        best_ambulance = ambulance
                    elif "Advanced Life Support" in best_ambulance_capabilities and "Advanced Life Support" not in ambulance_capabilities:
                        pass
                    elif ambulance["id"] < best_ambulance["id"]:
                        best_ambulance = ambulance
                elif ambulance["id"] < best_ambulance["id"]:
                    best_ambulance = ambulance

    # Step 2: Choose the best hospital
    best_hospital = None
    min_hospital_time = float('inf')

    if best_ambulance:
        for hospital in hospitals:
            hospital_region = hospital["region_id"]
            hospital_capabilities = hospital["capabilities"]
            emergency_capacity = hospital["emergency_capacity"]
            
            # Filter by matching capabilities
            hospital_match = False
            if emergency_type == "Cardiac" and "Cardiac Center" in hospital_capabilities:
                hospital_match = True
            elif emergency_type == "Trauma" and "Trauma Center" in hospital_capabilities:
                hospital_match = True
            elif emergency_type == "Stroke" and "Stroke Unit" in hospital_capabilities:
                hospital_match = True
            elif "General Emergency" in hospital_capabilities:
                hospital_match = True

            if not hospital_match and severity != "Critical":
                continue
            
            ambulance_to_hospital_time = dijkstra(graph, event_region, hospital_region, available_regions)

            if ambulance_to_hospital_time < min_hospital_time:
                min_hospital_time = ambulance_to_hospital_time
                best_hospital = hospital
            elif ambulance_to_hospital_time == min_hospital_time and best_hospital is not None:
                # Prioritize by emergency capacity (High > Medium > Low)
                capacity_priority = {"High": 3, "Medium": 2, "Low": 1}
                if capacity_priority[emergency_capacity] > capacity_priority[best_hospital["emergency_capacity"]]:
                    best_hospital = hospital

    result = {
        "emergency_id": emergency_id,
        "assigned_ambulance": best_ambulance["id"] if best_ambulance else None,
        "destination_hospital": best_hospital["id"] if best_hospital else None
    }

    return result

if __name__ == "__main__":
    input_json = json.load(input())
    emergency_event = input_json["emergency_event"]
    ambulances = input_json["ambulances"]
    hospitals = input_json["hospitals"]
    regions = input_json["regions"]
    
    result = dispatch_ambulance(emergency_event, ambulances, hospitals, regions)
    print(json.dumps(result))

File: codes/javascript/gemini/minimal/task_3/Solution.js
const process = require('process');

function dispatchAmbulance(emergency_event, ambulances, hospitals, regions) {
    const { id: emergency_id, region_id: emergency_region, severity, type, special_requirements } = emergency_event;

    // 2. Algorithm Implementation (Using Graph):
    //    a. Graph Construction:
    //       - Regions are nodes; distances form edges with weights calculated as travel times
    //       - Edge weight: Travel Time = Distance (km) / (Base Speed * Congestion Factor) * 60 minutes
    //         * Base Speed: Urban = 30 km/h, Suburban = 50 km/h, Rural = 80 km/h
    //         * Congestion Factor between 0-1 (1 = no congestion, 0.5 = 50% slower)
    //       - Assign infinite weight to blocked routes

    function calculateTravelTime(distance, regionType, congestionFactor) {
        let baseSpeed;
        switch (regionType) {
            case "Urban":
                baseSpeed = 30;
                break;
            case "Suburban":
                baseSpeed = 50;
                break;
            case "Rural":
                baseSpeed = 80;
                break;
            default:
                return Infinity; // Unknown region type
        }
        return distance / (baseSpeed * congestionFactor) * 60;
    }

    function buildGraph(regions) {
        const graph = {};
        regions.forEach(region => {
            graph[region.region_id] = {};
            for (const neighbor in region.distances) {
                if (region.blocked_routes.includes(neighbor)) {
                    graph[region.region_id][neighbor] = Infinity;
                } else {
                    graph[region.region_id][neighbor] = calculateTravelTime(
                        region.distances[neighbor],
                        region.type,
                        region.congestion_factor
                    );
                }
            }
        });
        return graph;
    }

    function dijkstra(graph, start, end) {
        const distances = {};
        const visited = {};
        const previous = {};
        for (const node in graph) {
            distances[node] = Infinity;
            visited[node] = false;
            previous[node] = null;
        }
        distances[start] = 0;

        while (true) {
            let closestNode = null;
            let shortestDistance = Infinity;
            for (const node in distances) {
                if (!visited[node] && distances[node] < shortestDistance) {
                    closestNode = node;
                    shortestDistance = distances[node];
                }
            }

            if (closestNode === null) {
                break;
            }

            visited[closestNode] = true;

            for (const neighbor in graph[closestNode]) {
                const distance = distances[closestNode] + graph[closestNode][neighbor];
                if (distance < distances[neighbor]) {
                    distances[neighbor] = distance;
                    previous[neighbor] = closestNode;
                }
            }
        }

        let path = [];
        let current = end;
        while (current !== null) {
            path.unshift(current);
            current = previous[current];
        }

        if (distances[end] === Infinity) {
            return { distance: Infinity, path: [] };
        }

        return { distance: distances[end], path: path };
    }

    const graph = buildGraph(regions);

    //    b. Two-Step Process (Fixed Order):
    //       - Step 1: Select the fastest ambulance that meets capability requirements using Dijkstra's algorithm
    //         * For tied arrival times, prioritize by capabilities, then lower ID
    //         * Critical emergencies require Advanced Life Support capabilities

    let bestAmbulance = null;
    let shortestAmbulanceTime = Infinity;

    for (const ambulance of ambulances) {
        if (ambulance.status === "Available") {
            let meetsRequirements = true;
            if (severity === "Critical" && special_requirements.includes("Advanced Life Support")) {
                if (!ambulance.capabilities.includes("Advanced Life Support") && !ambulance.capabilities.includes("Defibrillator")) {
                    meetsRequirements = false;
                }
            }

            if (meetsRequirements) {
                const ambulanceRoute = dijkstra(graph, ambulance.current_region, emergency_region);
                if (ambulanceRoute.distance < shortestAmbulanceTime) {
                    shortestAmbulanceTime = ambulanceRoute.distance;
                    bestAmbulance = ambulance;
                } else if (ambulanceRoute.distance === shortestAmbulanceTime) {
                    // Prioritize by capabilities, then lower ID
                    if (bestAmbulance && ambulance.capabilities.length > bestAmbulance.capabilities.length) {
                        bestAmbulance = ambulance;
                    } else if (bestAmbulance && ambulance.id < bestAmbulance.id) {
                        bestAmbulance = ambulance;
                    }
                }
            }
        }
    }

    if (!bestAmbulance) {
        return { emergency_id: emergency_id, assigned_ambulance: null, destination_hospital: null };
    }

    //       - Step 2: Using the selected ambulance, choose the best hospital using Dijkstra's algorithm
    //         * Filter by matching capabilities (Cardiac Center for cardiac cases, etc.)
    //         * Prioritize by emergency capacity (High > Medium > Low)
    //         * For Critical emergencies, prioritize response time over exact capability matching
    //         * Never reconsider ambulance selection even if no perfect hospital match exists

    let bestHospital = null;
    let shortestHospitalTime = Infinity;

    for (const hospital of hospitals) {
        let meetsRequirements = true;

        if (type === "Cardiac" && !hospital.capabilities.includes("Cardiac Center")) {
            meetsRequirements = false;
        }

        if (meetsRequirements) {
            const hospitalRoute = dijkstra(graph, emergency_region, hospital.region_id);
            if (hospitalRoute.distance < shortestHospitalTime) {
                shortestHospitalTime = hospitalRoute.distance;
                bestHospital = hospital;
            } else if (hospitalRoute.distance === shortestHospitalTime) {
                // Prioritize by emergency capacity (High > Medium > Low)
                if (bestHospital) {
                    const capacityOrder = { "High": 3, "Medium": 2, "Low": 1 };
                    if (capacityOrder[hospital.emergency_capacity] > capacityOrder[bestHospital.emergency_capacity]) {
                        bestHospital = hospital;
                    }
                }
            }
        }
    }

    if (severity === "Critical" && !bestHospital) {
      for (const hospital of hospitals) {
        const hospitalRoute = dijkstra(graph, emergency_region, hospital.region_id);
        if (hospitalRoute.distance < shortestHospitalTime) {
            shortestHospitalTime = hospitalRoute.distance;
            bestHospital = hospital;
        } else if (hospitalRoute.distance === shortestHospitalTime) {
            // Prioritize by emergency capacity (High > Medium > Low)
            if (bestHospital) {
                const capacityOrder = { "High": 3, "Medium": 2, "Low": 1 };
                if (capacityOrder[hospital.emergency_capacity] > capacityOrder[bestHospital.emergency_capacity]) {
                    bestHospital = hospital;
                }
            }
        }
      }
    }

    return {
        emergency_id: emergency_id,
        assigned_ambulance: bestAmbulance ? bestAmbulance.id : null,
        destination_hospital: bestHospital ? bestHospital.id : null
    };
}

// Input/Output Requirements:
//    a. Do NOT modify the input data structures
//    b. Always return a complete object with: emergency_id, assigned_ambulance, destination_hospital
//    c. When no assignment is possible, return fields as null but maintain the object structure
//    d. Follow the exact output format shown in the examples
process.stdin.setEncoding('utf8');

let input = '';

process.stdin.on('data', function(chunk) {
    input += chunk;
});

process.stdin.on('end', function() {
    try {
        const data = JSON.parse(input);
        const result = dispatchAmbulance(data.emergency_event, data.ambulances, data.hospitals, data.regions);
        console.log(JSON.stringify(result));
    } catch (error) {
        console.error("Error processing input:", error);
    }
});